codalab/worker/docker_image_manager.py

```diff
@@ -218,12 +224,34 @@ def download():
                             'message'
                         ] = "Can't download image: {}".format(ex)
 
-                self._downloading.add_if_new(image_spec, threading.Thread(target=download, args=[]))
+                # Check docker image size before pulling from docker hub.
+                # Images with size larger than '10g' will be rejected to download.
+                image_size_bytes = docker_utils.get_image_size_without_pulling(image_spec)
+                if 0 < image_size_bytes < self._max_image_size:
+                    self._downloading.add_if_new(
+                        image_spec, threading.Thread(target=download, args=[])
+                    )
+                    return ImageAvailabilityState(
+                        digest=None,
+                        stage=DependencyStage.DOWNLOADING,
+                        message=self._downloading[image_spec]['status'],
+                    )
+                elif image_size_bytes == 0:
```

Comment:
Don't use `0` to denote failure since it's too cryptic / easy to miss that case....I'd throw an exception or return `None`.